cancer_ai/validator/model_db.py

```diff
@@ -0,0 +1,184 @@
+            if block_hash is None:
+                raise ValueError(f"Block hash not found for block number {block_number}")
+            
+            timestamp_info = self.subtensor.substrate.query(
+                module='Timestamp',
+                storage_function='Now',
+                block_hash=block_hash
+            )
+
+            if timestamp_info is None:
+                raise ValueError(f"Timestamp not found for block hash {block_hash}")
+
+            timestamp_ms = timestamp_info.value
+            block_datetime = datetime.fromtimestamp(timestamp_ms / 1000.0)
+
+            return block_datetime
+        except Exception as e:
+            bt.logging.error(f"Error retrieving block timestamp: {e}")
+            raise
+
+    def get_latest_models(self, hotkeys: list[str]) -> list[ChainMinerModel]:
+        # Set the cutoff time to 30 minutes ago
+        cutoff_time = datetime.now() - timedelta(minutes=RETRIEVE_MODELS_TIME_FRAME)
+
+        session = self.Session()
+        try:
+            # Use a correlated subquery to get the latest record for each hotkey that doesn't violate the cutoff
+            latest_models = []
+            for hotkey in hotkeys:
+                model_record = (
+                    session.query(ChainMinerModelDB)
+                    .filter(ChainMinerModelDB.hotkey == hotkey)
+                    .filter(ChainMinerModelDB.date_submitted < cutoff_time)
+                    .order_by(ChainMinerModelDB.date_submitted.desc())  # Order by newest first
+                    .first()  # Get the first (newest) record that meets the cutoff condition
+                )
+                if model_record:
+                    latest_models.append(
+                        ChainMinerModel(
+                            competition_id=model_record.competition_id,
+                            hf_repo_id=model_record.hf_repo_id,
+                            hf_model_filename=model_record.hf_model_filename,
+                            hf_repo_type=model_record.hf_repo_type,
+                            hf_code_filename=model_record.hf_code_filename,
+                            block=model_record.block,
+                            hotkey=model_record.hotkey,
+                        )
+                    )
+
+            return latest_models
+        finally:
+            session.close()
+
+    def clean_old_records(self, hotkeys: list[str]):
+        session = self.Session()
+        try:
+            for hotkey in hotkeys:
+                # Query all records for this hotkey, ordered by date_submitted in descending order
+                records = (
+                    session.query(ChainMinerModelDB)
+                    .filter(ChainMinerModelDB.hotkey == hotkey)
+                    .order_by(ChainMinerModelDB.date_submitted.desc())
+                    .all()
+                )
+
+                # If there are more than STORED_MODELS_PER_HOTKEY records, delete the oldest ones
```

Comment:
it could be done directly in `add_model`, no need for cleaning old records

Reply:
Id rather have such functionality in the method which is called clean_old_records rather then add_model
It makes more sense to me that way